lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -106,24 +113,90 @@ bool SessionStateEnforcer::m5g_init_session_credit(
   return true;
 }
 
-void SessionStateEnforcer::handle_session_init_rule_updates(
+bool SessionStateEnforcer::handle_session_init_rule_updates(
     SessionMap& session_map, const std::string& imsi,
-    SessionState& session_state) {
-  uint32_t l_teid = 0;
-  auto itp        = pdr_map_.equal_range(imsi);
+    std::unique_ptr<SessionState>& session_state) {
+  uint32_t upf_teid = 0;
+  uint32_t len, gnode_teid = 0;
+  char ip_str[INET_ADDRSTRLEN];
+  struct sockaddr_in ip_sa;
+  bool byte_zero_skip, upf_teid_set = false;
+  std::string ip_addr, teid;
+  SetGroupPDR rule;
+
+  const auto& config = session_state->get_config();
+  /*
+   * Lets add the default uplink and downlink PDR rule to this
+   * imsi
+   */
+  pdr_map_.insert(
+      std::pair<std::string, uint32_t>(imsi, DEFAULT_DOWN_LINK_PDR_ID));
+  pdr_map_.insert(
+      std::pair<std::string, uint32_t>(imsi, DEFAULT_UP_LINK_PDR_ID));
+  /*
+   * Plus policy DB added anymore it would be part of
+   * this map
+   */
+  auto itp = pdr_map_.equal_range(imsi);
+  // Lets take local_teid of the session if already exists.
+  upf_teid = session_state->get_local_teid();
   for (auto itr = itp.first; itr != itp.second; itr++) {
     // Get the PDR numbers, now  get the rules from global static rule list
-    SetGroupPDR rule;
     GlobalRuleList.get_rule(itr->second, &rule);
-    session_state.insert_pdr(&rule);
-    // Get new TEID
-    if (!l_teid) l_teid = pipelined_client_->get_current_teid();
-    rule.mutable_pdi()->set_local_f_teid(pipelined_client_->get_next_teid());
+    // Check session pdr list, if its already part of the list
+    if (session_state->search_pdr(rule.pdr_id()))
+      continue;  // If rule alreay added part of Session PdrList
+    // Get the UE ip address
+    ip_addr.assign(config.rat_specific_context.m5gsm_session_context()
+                       .pdu_address()
+                       .redirect_server_address());
+    memcpy(&(ip_sa.sin_addr), ip_addr.c_str(), ip_addr.length());
+    inet_ntop(AF_INET, &(ip_sa.sin_addr), ip_str, INET_ADDRSTRLEN);
+    rule.mutable_pdi()->set_ue_ip_adr(ip_str);
+    switch (rule.pdi().src_interface()) {
+      case ACCESS:
+        // Get new UPF TEID

```

Comment:
Can we define separate functions for inserting rules from different interfaces here?
Something like:
```
switch (rule.pdi().src_interface()) {
      case ACCESS:
        insert_pdr_from_access(args...);
        break;
      case CORE:
        insert_pdr_from_core(args...);
        break;
    }
```

Reply:
Good suggesion, adopted GNB endpoint updating.

Reply:
Adopted.